extensions/amp-ima-video/0.1/amp-ima-video.js

```diff
@@ -0,0 +1,224 @@
+import {assertHttpsUrl} from '../../../src/url';
+import {getIframe, preloadBootstrap} from '../../../src/3p-frame';
+import {
+  installVideoManagerForDoc,
+} from '../../../src/service/video-manager-impl';
+import {isExperimentOn} from '../../../src/experiments';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {isObject} from '../../../src/types';
+import {
+  listen,
+  loadPromise,
+} from '../../../src/event-helper';
+import {removeElement} from '../../../src/dom';
+import {user} from '../../../src/log';
+import {VideoEvents} from '../../../src/video-interface';
+import {videoManagerForDoc} from '../../../src/services';
+
+/** @const */
+const TAG = 'amp-ima-video';
+
+/**
+ * @implements {../../../src/video-interface.VideoInterface}
+ */
+class AmpImaVideo extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.playerReadyPromise_ = null;
+
+    /** @private {?Function} */
+    this.playerReadyResolver_ = null;
+
+    /** @private {?Function} */
+    this.unlistenMessage_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    user().assert(isExperimentOn(this.win, TAG),
+        'Experiment ' + TAG + ' is disabled.');
+
+    assertHttpsUrl(this.element.getAttribute('data-tag'),
+        'The data-tag attribute is required for <amp-video-ima> and must be ' +
+            'https',
+        this.element);
+    assertHttpsUrl(this.element.getAttribute('data-src'),
+        'The data-src attribute is required for <amp-video-ima> and must be ' +
+            'https',
+        this.element);
+  }
+
+  /** @override */
+  preconnectCallback() {
+    this.preconnect.preload(
+        'https://imasdk.googleapis.com/js/sdkloader/ima3.js', 'script');
+    preloadBootstrap(this.win, this.preconnect);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  layoutCallback() {
+    const iframe = getIframe(this.element.ownerDocument.defaultView,
+        this.element, 'ima-video');
+    iframe.setAttribute('allowfullscreen', 'true');
+    this.applyFillContent(iframe);
+    this.element.appendChild(iframe);
```

Comment:
please move `appendChild()` to be after `listen` (to avoid a rare race condition)